frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/GitFetchSupportImpl.java

```diff
@@ -0,0 +1,462 @@
+package com.virtuslab.gitmachete.frontend.actions.common;
+
+import static com.intellij.openapi.vcs.VcsNotifier.STANDARD_NOTIFICATION;
+import static git4idea.GitUtil.findRemoteByName;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import com.intellij.dvcs.MultiMessage;
+import com.intellij.dvcs.MultiRootMessage;
+import com.intellij.internal.statistic.IdeActivity;
+import com.intellij.notification.NotificationType;
+import com.intellij.openapi.progress.EmptyProgressIndicator;
+import com.intellij.openapi.progress.ProcessCanceledException;
+import com.intellij.openapi.progress.ProgressManager;
+import com.intellij.openapi.progress.util.BackgroundTaskUtil;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.util.registry.Registry;
+import com.intellij.openapi.vcs.VcsException;
+import com.intellij.openapi.vcs.VcsNotifier;
+import com.intellij.util.concurrency.AppExecutorUtil;
+import git4idea.GitUtil;
+import git4idea.commands.Git;
+import git4idea.commands.GitAuthenticationGate;
+import git4idea.commands.GitAuthenticationListener;
+import git4idea.commands.GitImpl;
+import git4idea.commands.GitRestrictingAuthenticationGate;
+import git4idea.config.GitConfigUtil;
+import git4idea.fetch.GitFetchResult;
+import git4idea.fetch.GitFetchSupport;
+import git4idea.repo.GitRemote;
+import git4idea.repo.GitRepository;
+import io.vavr.control.Option;
+import io.vavr.control.Try;
+import kotlin.jvm.functions.Function1;
+import lombok.AllArgsConstructor;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import com.virtuslab.gitmachete.frontend.actions.toolbar.FetchAllRemotesAction;
+import com.virtuslab.logger.IPrefixedLambdaLogger;
+import com.virtuslab.logger.PrefixedLambdaLoggerFactory;
+
+/**
+ * @deprecated This implementation is a workaround to provide features missing in pre-2020.1 versions.
+ * The main benefit of it is specific branch update ({@link GitFetchSupportImpl#fetch(GitRepository, GitRemote, String)}).
+ * We need this class for {@link FetchAllRemotesAction} and {@link BasePullBranchAction}.
+ */
+@Deprecated
+public final class GitFetchSupportImpl implements GitFetchSupport {
```

Comment:
Just for clarity - how much of that class is a port of the existing `GitFetchSupportImpl` from Kotlin to Java, and how much is bespoke implementation for our plugin?

Reply:
We need implementation of two methods:
- `GitFetchSupportImpl#fetchAllRemotes`
- `GitFetchSupportImpl#fetch(git4idea.repo.GitRepository, git4idea.repo.GitRemote, java.lang.String)` (this one belongs to the interface since 201)

To be honest a simple specified `refspec` handling introduced a lot of changes to the class. Adding `RemoteRefCoordinates` dto is an example of the impact.

Reply:
So this is what happens:
`15:58:22.909: [filter-service] git -c credential.helper= -c core.quotepath=false -c log.showSignature=false fetch origin develop:develop --recurse-submodules=no --progress --prune`
![image](https://user-images.githubusercontent.com/19799111/81829404-2388d300-953b-11ea-8e4a-18f0cfc63701.png)